components/script/dom/gpubuffer.rs

```diff
@@ -30,31 +33,96 @@ use crate::dom::promise::Promise;
-const RANGE_SIZE_ALIGN_MASK: u64 = 4;
-
-// https://gpuweb.github.io/gpuweb/#buffer-state
-#[derive(Clone, Copy, MallocSizeOf, PartialEq)]
-pub enum GPUBufferState {
-    Mapped,
-    MappedAtCreation,
-    MappingPending,
-    Unmapped,
-    Destroyed,
-}
-
 #[derive(JSTraceable, MallocSizeOf)]
-pub struct GPUBufferMapInfo {
-    #[ignore_malloc_size_of = "Arc"]
+pub struct MappingView {
     #[no_trace]
-    /// The `mapping` is wrapped in an `Arc` to ensure thread safety.
-    /// This is necessary for integration with the SpiderMonkey engine,
-    pub mapping: Arc<Mutex<Vec<u8>>>,
-    pub mapping_range: Range<u64>,
-    pub mapped_ranges: Vec<Range<u64>>,
+    range: Range<usize>,
     #[ignore_malloc_size_of = "defined in mozjs"]
-    pub js_buffers: Vec<HeapBufferSource<ArrayBufferU8>>,
-    pub map_mode: Option<u32>,
+    buffer: HeapArrayBuffer,
+}
+
+impl Drop for MappingView {
+    #[allow(unsafe_code)]
+    fn drop(&mut self) {
+        let cx = GlobalScope::get_cx();
+        assert!(unsafe { DetachArrayBuffer(*cx, self.buffer.underlying_object().handle()) })
+    }
+}
+
+#[derive(JSTraceable, MallocSizeOf)]
+pub struct ActiveBufferMapping {
+    // TODO(sagudev): Use IpcSharedMemory when https://github.com/servo/ipc-channel/pull/356 lands
+    /// <https://gpuweb.github.io/gpuweb/#active-buffer-mapping-data>
+    #[ignore_malloc_size_of = "Rc"]
+    data: Arc<Box<[u8]>>,
+    /// <https://gpuweb.github.io/gpuweb/#active-buffer-mapping-mode>
+    mode: GPUMapModeFlags,
+    /// <https://gpuweb.github.io/gpuweb/#active-buffer-mapping-range>
+    range: Range<u64>,
+    /// <https://gpuweb.github.io/gpuweb/#active-buffer-mapping-views>
+    views: Vec<MappingView>,
+}
+
+impl ActiveBufferMapping {
+    /// <https://gpuweb.github.io/gpuweb/#abstract-opdef-initialize-an-active-buffer-mapping>
+    pub fn new(mode: GPUMapModeFlags, range: Range<u64>) -> Fallible<Self> {
+        // Step 1
+        let size: usize = (range.end - range.start).try_into().unwrap();
+        // Step 2
+        if size > (1 << 53) - 1 {
+            return Err(Error::Range("Over MAX_SAFE_INTEGER".to_string()));
+        }
+        let data = vec![0; size];
+        Ok(Self {
+            data: Arc::new(data.into_boxed_slice()),
+            mode,
+            range,
+            views: Vec::new(),
+        })
+    }
+
+    /// Panics if there is any active view
+    pub fn load_data(&mut self, src: &[u8]) {
+        // `Arc::get_mut` ensures there are no views
+        Arc::get_mut(&mut self.data).unwrap().clone_from_slice(src)
+    }
+
+    #[allow(unsafe_code)]
+    pub fn view(&mut self, range: Range<usize>) -> Result<&MappingView, ()> {
+        if self
+            .views
+            .iter()
+            .any(|view| range_overlap(&view.range, &range))
+        {
+            return Err(());
+        }
+        let cx = GlobalScope::get_cx();
+        /// `freeFunc()` must be threadsafe, should be safely callable from any thread
```

Comment:
Is there a reason why you didn't just change the signature of `create_new_external_array_buffer` to take an `Arc<Box<[u8]>>`, and store a `HeapBufferSource` as the buffer of `MappingView`? The goal of `create_new_external_array_buffer` was to move the unsafe code into `bindings`, as opposed to do it in the dom object directly. Would there be a way to achieve your goal while maintaining that boundary?

Reply:
create_new_external_array_buffer was actually unsafe, as one can create multiple mut views to same slice, for safety we must pack this in `ActiveBufferMapping` to make sure there is no overlap. We could still have `create_new_external_array_buffer` abstracted away, but I think that would make all safety invariants less clear.

Reply:
Ok I see, then perhaps move `ActiveBufferMapping` to `bindings`? It may be useful elsewhere than only gpubuffer?

Reply:
Not really, as this is from webgpu spec https://gpuweb.github.io/gpuweb/#active-buffer-mapping

Reply:
Ok thanks for pointing it out. I see that the unsafe code seems to relate to https://gpuweb.github.io/gpuweb/#active-buffer-mapping-views, which is a [list](https://infra.spec.whatwg.org/#list)<[ArrayBuffer](https://webidl.spec.whatwg.org/#idl-ArrayBuffer)> so not specific to WebGPU, So perhaps you can add a struct in bindings that encaspulates the idea of a safely managed list of views into a mapping? For example something like a `MappingManager`, with a safe `view` method returning a `MappingView`(which I think could be moved to bindings as well).

Reply:
Done in https://github.com/servo/servo/pull/33154/commits/f792055b697212776611b9d5a2da1b45dc0ee560 (with more generic names).

Reply:
thank you! (I was just gonna say this could be a good follow-up too...)